Add a stack variable's align() to the AST

JohanEngelen opened the pull request:
This PR is to support code such as:

``` D
void f() {
    align(16) int a; // <---- this alignment is now stored in the AST
}
```

The align(16) is parsed at line 3961, however it is never applied to the Dsymbol for stack variables (line 4022 and onwards). It is simply ignored, but it should be stored in the AST for the backend to use (DMD backend does not use it yet).
I am not at all familiar with the parser code, thanks for the kind review.

The bug was found in LDC's C-version of DMD, see https://github.com/ldc-developers/ldc/pull/1214. With the changes of this PR, LDC is able to apply `align()` to stack variables correctly.

Commit:
Add a stack variable's alignment to the AST.

This is to support code such as:
void f() {
	align(16) int a; // <---- this alignment is now stored in the AST
}

The align(16) is parsed at line 3961, however it is never applied to the Dsymbol for stack variables (line 4022 and onwards). It should be stored in the AST for the backend to use (DMD backend does not use it yet).

ibuclaw commented on src/parse.d:
```diff
@@ -4087,10 +4087,16 @@ public:
                     pAttrs.storageClass &= (STCsafe | STCsystem | STCtrusted);
                 }
                 Dsymbol s = v;
+                if (structalign != 0)
```
Wasn't there an enum value for no struct alignment set?

kinke replied:
In the C version, there's a `STRUCTALIGN_DEFAULT = ~0` #define.

ibuclaw replied:
Right, so as the value of structalign should have been assigned correctly [here](https://github.com/D-Programming-Language/dmd/pull/5284/files#diff-e55efe6f4373bd0f2e2da9a75322a544R3762), shouldn't this check be ``structalign != STRUCTALIGN_DEFAULT`` ?

ibuclaw replied:
Oh, this is a third meaning given to the alignment value (``>0``, ``~0`` and ``0``)

JohanEngelen replied:
Indeed. Note that the code is identical to line 3973.

JohanEngelen replied:
@ibuclaw @kinke Note that the code is identical to line 3973.

JohanEngelen commented:
I have removed some of the changes. Now it only adds alignment to the parse tree for
```D
align(16) int a;
```


More work (beyond the parser) is needed for 
```D
alias Integer16 = align(16) int;
Integer16 i16; // (16-byte aligned)
```
and so I have removed the partial changes towards that goal, because I cannot test it (the changes were in the wrong place actually). I can add these changes back in, if you think it is better (this time in the correct place).

WalterBright commented:
Auto-merge toggled on

ibuclaw commented:
> Now it only adds alignment to the parse tree for

No compilable testcase?

JohanEngelen commented:
@ibuclaw  You mean adding a testcase to test that the parser accepts `align(16) int a;` ?

A runnable testcase would be (once DMD implements stack variable alignment):
```D
void main()
{
    byte dummy;

    align(32) int align32;
    assert((cast(size_t)&align32 & cast(size_t)0b11111) == 0);
}
```

ibuclaw commented:
> You mean adding a testcase to test that the parser accepts align(16) int a;

Yes.  Don't need to worry about a runnable test.
